lib/charms/mlops_libs/v0/k8s_service_info.py

```diff
@@ -0,0 +1,297 @@
+        self.message = message
+        super().__init__(self.message)
+
+
+class KubernetesServiceInfoRequirer(Object):
+    """Base class that represents a requirer relation end.
+
+    Args:
+        requirer_charm (CharmBase): the requirer application
+        relation_name (str, optional): the name of the relation
+
+    Attributes:
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    def __init__(self, requirer_charm, relation_name: str = DEFAULT_RELATION_NAME):
+        super().__init__(requirer_charm, relation_name)
+        self.relation_name = relation_name
+
+    @staticmethod
+    def _relation_preflight_checks(relation: Relation) -> None:
+        """Series of checks for the relation and relation data.
+
+        Args:
+            relation (Relation): the relation object to run the checks on
+
+        Raises:
+            KubernetesServiceInfoRelationDataMissingError if data is missing or incomplete
+            KubernetesServiceInfoRelationMissingError: if there is no related application
+            ops.model.TooManyRelatedAppsError: if there is more than one related application
+        """
+        # Raise if there is no related application
+        if not relation:
+            raise KubernetesServiceInfoRelationMissingError()
+
+        # Extract remote app information from relation
+        remote_app = relation.app
+        # Get relation data from remote app
+        relation_data = relation.data[remote_app]
+
+        # Raise if there is no data found in the relation data bag
+        if not relation_data:
+            raise KubernetesServiceInfoRelationDataMissingError(
+                f"No data found in relation {relation.name} data bag."
+            )
+
+        # Check if the relation data contains the expected attributes
+        missing_attributes = [
+            attribute for attribute in REQUIRED_ATTRIBUTES if attribute not in relation_data
+        ]
+        if missing_attributes:
+            raise KubernetesServiceInfoRelationDataMissingError(
+                f"Missing attributes: {missing_attributes} in relation {relation.name}"
+            )
+
+    def get_k8s_svc_info(self) -> dict:
+        """Return a dictionary with the Kubernetes Service information.
+
+        Raises:
+            KubernetesServiceInfoRelationDataMissingError: if data is missing entirely or some attributes
+            KubernetesServiceInfoRelationMissingError: if there is no related application
+        """
+        # Run pre-flight checks
+        # Raises TooManyRelatedAppsError if related to more than one app
+        relation = self.model.get_relation(self.relation_name)
+        self._relation_preflight_checks(relation=relation)
+
+        # Get relation data from remote app
+        relation_data = relation.data[relation.app]
+
+        return {
+            "svc_name": relation_data["svc_name"],
+            "svc_port": relation_data["svc_port"],
+        }
```

Comment:
Not a hard requirement, but consider returning an object with `.name` and `.port` here instead of an unstructured dict.  The dict means nobody downstream can have type hints or advanced linting.  

I did something like that in [kubeflow-dashboards](https://github.com/canonical/kubeflow-dashboard-operator/blob/05ac84bf63bff38da543f2f3a77a8123966e1d3f/lib/charms/kubeflow_dashboard/v0/kubeflow_dashboard_links.py#L176).  I think if I understood pydantic better at the time, I could have even used the same [pydantic object](https://github.com/canonical/charm-relation-interfaces/blob/a26d4550e72c751967117abf31b032cee5e1a0dc/interfaces/kubeflow_dashboard_links/v0/schema.py#L26).